contracts/c-sphincs-all-in-one-lock/ckb-sphincsplus-root-lock.c

```diff
@@ -0,0 +1,406 @@
+        ERROR_SPHINCSPLUS_ENCODING);
+
+    /*
+     * Iterate over lock data again to verify each signature.
+     */
+    while (signatures.size > 0) {
+      uint8_t header;
+      CHECK(mol2_read_and_advance(&signatures, &header, 1));
+
+      uint8_t param_id = header & MULTISIG_PARAMS_ID_MASK;
+      if (all_params[param_id] == NULL) {
+        CHECK(fetch_params(param_id, &all_params[param_id]));
+      }
+      uint32_t pk_size = all_params[param_id]->pk_bytes;
+      uint32_t sign_size = all_params[param_id]->sign_bytes;
+      CHECK2(pk_size + sign_size + 1 == (size_t)pk_size + (size_t)sign_size + 1,
+             ERROR_SPHINCSPLUS_WITNESS;);
+
+      if ((header & MULTISIG_SIG_MASK) != 0) {
+        /*
+         * Spawns a leaf process if needed, and use it to validate
+         * the signature
+         */
+        CHECK2(signatures.size >= pk_size + sign_size,
+               ERROR_SPHINCSPLUS_WITNESS);
+        /* Spawn a new proces if needed */
+        if (spawned_processes[param_id] == 0) {
+          uint64_t root_to_leaf_fds[2];
+          CHECK(ckb_pipe(root_to_leaf_fds));
+          uint64_t leaf_to_root_fds[2];
+          CHECK(ckb_pipe(leaf_to_root_fds));
+
+          /*
+           * Leaf process would read from root_to_leaf_fds[0],
+           * and write to leaf_to_root_fds[1]
+           */
+          uint64_t inherited_fds[3] = {root_to_leaf_fds[0], leaf_to_root_fds[1],
+                                       0};
+          fds[param_id][0] = root_to_leaf_fds[1];
+          fds[param_id][1] = leaf_to_root_fds[0];
+
+          const char *argv[] = {(const char *)escaped};
+          spawn_args_t spawn_args = {.argc = 1,
+                                     .argv = (const char **)argv,
+                                     .process_id = &spawned_processes[param_id],
+                                     .inherited_fds = inherited_fds};
+          CHECK(ckb_spawn_cell(code_hash, hash_type,
+                               *all_params[param_id]->offset_ptr,
+                               *all_params[param_id]->length_ptr, &spawn_args));
+        }
+        /* Sends data location to the leaf process */
+        {
+          uint8_t data[3 + 8 + 4 + 4 + 4];
+          /* The first 3 bytes are ckb-script-ipc compatible headers */
+          data[0] = 0;  /* Version in VLQ encoding */
+          data[1] = 1;  /* Method ID(1) in VLQ encoding */
+          data[2] = 20; /* Length of Payload(20) in VLQ encoding */
+          *((uint64_t *)&data[3]) = CKB_SOURCE_GROUP_INPUT;
+          *((uint64_t *)&data[3 + 8]) = 0;
+          *((uint64_t *)&data[3 + 8 + 4]) = signatures.offset - 1;
+          *((uint64_t *)&data[3 + 8 + 4 + 4]) = 1 + pk_size + sign_size;
+
+          CHECK(_write_all(fds[param_id][0], data, sizeof(data)));
+        }
+        /* Waits for leaf process to respond */
+        {
+          uint8_t response;
+          CHECK(_read_all(fds[param_id][1], &response, 1));
```

Comment:
This simplified version of `receive Response` works if the callee writes the Request using a single `ckb_write` syscall.  
However, if the request is split across multiple write syscalls, it fails.  
The correct approach is to first retrieve the payload size from the header, then use `read_exact` to read the remaining data.